### PR TITLE
Add acl api

### LINKED ISSUE #5296: ACL backend support
DoD:

- [x] lakeFS server-side support for enforcing ACLs.
- [x] lakeFS server-side support for editing ACLs.

### PR DESCRIPTION
**Pull request by arielshaqed**
### Linked Issue

Closes #5296 

---

### Background

Add backend support for ACLs and remove API support for modifying policies.  Part of implementing [this security update](https://docs.lakefs.io/posts/security_update.html).
      
### New Feature

Backend support for lakeFS "Simplified" authorization mode.  Add support and API for ACLs.  In parallel, remove RBAC from lakeFS backend.  Also add lakectl support for ACLs -- helps validation and manual testing, and will be a part of the UI once we unhide it.

⚠️ This is not yet a complete implementation of simplified mode!  Missing are:
* [x] ⚠️ GUI
* [x] ⚠️ Upgrade script
We will not pull before all of these are ready.

### Testing Details

Added unit tests for ACL enforcement.  Note that previously we had no direct testing of P/RBAC enforcement, this is actually an improvement in terms of direct testing support.

Manual testing with lakectl:
* [x] CRUD operations.
* [x] End-to-end enforcement of actual generated policies.

### Breaking Change?

😈 Hell yes! users with existing P/RBAC configurations will have their configurations changed by the upcoming upgrade script.  Some previously-enforceable policies will now require the use of an external user-provided authorization server to implement.  See [the design](#5267) for details.

### CODE REVIEW

**Comment by arielshaqed** [comment]
Reviewers: Please do not be alarmed by the immensity of this PR.  As large as it is, a even larger proportion of it is due to API generated code.  Ignore everything under clients/, it's all generated code.

**Comment by arielshaqed** [comment]
## Manual test: Generated policy is attached to user

I created a user `noob` in a group `noobs`.

```
❯ lakectl auth groups acl set --id noobs --permission Read --repositories foo

# (separator lines added by ariels)

❯ lakectl -c ~/.lakectl.s3.yaml auth groups acl get --id noobs
Group noobs:Read
+------------+
| REPOSITORY |
+------------+
| foo        |
+------------+


This user has read-only access to the single repository "foo":
```
❯ lakectl -c ~/.lakectl.noob.yaml repo list
insufficient permissions
401 Unauthorized
❯ lakectl -c ~/.lakectl.noob.yaml fs ls lakefs://foo/
Invalid 'path': not a valid path uri
Error executing command.
❯ lakectl -c ~/.lakectl.noob.yaml fs ls lakefs://foo/main/
object          2022-09-21 11:49:17 +0300 IDT    4 B             moo
common_prefix                                                    oh/
❯ lakectl -c ~/.lakectl.noob.yaml fs ls lakefs://bar/main/
insufficient permissions
401 Unauthorized
❯ echo nope | lakectl -c ~/.lakectl.noob.yaml fs upload lakefs://foo/main/nope -s - 
insufficient permissions
401 Unauthorized
```

File: pkg/auth/acl/permission.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+func RepositoriesToARNs(repositories model.Repositories) []string {
+	if repositories.All {
+		// "*" always treated globally, no specific ARN.
+		return all
+	}
+	arns := make([]string, len(repositories.List))
+	for i, repo := range repositories.List {
+		arns[i] = permissions.RepoArn(repo)
+	}
+	return arns
```
Comment (guy-har) [inline pkg/auth/acl/permission.go:45]: This looks fine to me, but just in order to validate, 
`!repositories.All` => `len(repositories.List) !=0`
Is that the case? or is empty string a valid response?

File: pkg/auth/setup.go [hunk 2]
```diff
@@ -47,6 +52,85 @@ func attachPolicies(ctx context.Context, authService Service, groupID string, po
...
+// MakeStatementForPolicyType returns statements for policy type typ,
+// limited to resources.
+func MakeStatementForPolicyType(typ string, resources []string) (model.Statements, error) {
+	statement, ok := statementByName[typ]
+	if !ok {
+		return nil, fmt.Errorf("%w: %s", ErrStatementNotFound, typ)
+	}
+	statements := make(model.Statements, len(resources))
+	for i, resource := range resources {
+		if statements[i].Resource == "" {
```
Comment (guy-har) [inline pkg/auth/setup.go:118]: isn't this `if` redundant?

File: pkg/auth/setup.go [hunk 3]
```diff
@@ -47,6 +52,85 @@ func attachPolicies(ctx context.Context, authService Service, groupID string, po
...
+			statements[i] = statement
+			statements[i].Resource = resource
+		}
+	}
+	return statements, nil
+}
+
+func MakeStatementForPolicyTypeOrDie(typ string, resources []string) model.Statements {
```
Comment (guy-har) [inline pkg/auth/setup.go:126]: Suggestion: change `typ` from string to type

**Review by guy-har — APPROVED** [review]
Looks great.
Added some comments, non of them are blocking

File: pkg/auth/model/model.proto [hunk 4]
```diff
@@ -50,3 +59,14 @@ message TokenData {
...
+message RepositoriesData {
+    bool all = 1;
+    repeated string list = 2;
+}
+
+// message data model for UI structure
+message UIData {
+    string permission = 1;
+    RepositoriesData repositories = 2;
+}
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/auth/model/model.proto:72]: Completeness.  But also consider a Cloud user who has to use external authentication but wants simplified mode.

File: pkg/auth/service.go [hunk 5]
```diff
@@ -738,6 +738,9 @@ func (s *AuthService) WritePolicy(ctx context.Context, policy *model.Policy, upd
 
 	if update {
 		_, err := kv.GetMsg(ctx, s.store, model.PartitionKey, policyKey, &model.PolicyData{})
+		if errors.Is(err, kv.ErrPredicateFailed) {
+			err = ErrNotFound
+		}
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply pkg/auth/service.go]: You're right.  Bad copy-pasta.  Removed.

File: pkg/auth/setup.go (hunk 2 above)
Comment (arielshaqed) [reply pkg/auth/setup.go]: True.  It should look at statement, which _might_ in future be configured with a specific resource.  It was meant as a safety measure for when that day comes, so we do not end up _widening_ a limited statement in `statementByName`.

File: pkg/auth/setup.go (hunk 3 above)
Comment (arielshaqed) [reply pkg/auth/setup.go:126]: Not done: this is a word in the GUI.  In general, the raw string / named string type distinction in package auth is unclear.

**Review by arielshaqed — COMMENTED** [review]
THANKS!

**Comment by arielshaqed** [comment]
Marking as "draft":
* Merging is allowed!
* But without the followup PRs (into this branch) merging would leave an unusable lakeFS

**Comment by arielshaqed** [comment]
This is well and truly ready.  @itaiad200 do you want anyone from VE to look at this?  So far no traction.  Apologies for the large PR, it's been growing for a while.

Most important changes for VE are probably up to 06eebafb88e13e3c2fe4cf6c29cdfffa1a0b112a.

**Comment by arielshaqed** [comment]
Rebased (conflicts with trunk were just auto-generated files).  Waiting for tests to pass...

**Pull request merged**